lib-multisrc/libgroup/src/eu/kanade/tachiyomi/multisrc/libgroup/LibGroup.kt

```diff
@@ -18,284 +23,203 @@ import eu.kanade.tachiyomi.source.model.Page
     }
 
-    private fun imgHeader() = Headers.Builder().apply {
-        add("User-Agent", userAgentMobile)
-        add("Accept", "image/avif,image/webp,*/*")
-        add("Referer", baseUrl)
-    }.build()
-
-    protected fun catalogHeaders() = Headers.Builder()
-        .apply {
-            add("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.127 Safari/537.36 Edg/100.0.$userAgentRandomizer")
-            add("Accept", "application/json, text/plain, */*")
-            add("X-Requested-With", "XMLHttpRequest")
-            add("x-csrf-token", csrfToken)
+    private var _constants: Constants? = null
+    private fun getConstants(): Constants {
+        if (_constants == null) {
+            try {
+                _constants = client.newCall(
+                    GET("https://api.$apiDomain/api/constants?fields[]=genres&fields[]=tags&fields[]=types&fields[]=scanlateStatus&fields[]=status&fields[]=format&fields[]=ageRestriction&fields[]=imageServers", headersBuilder().build()),
+                ).execute().parseAs<Data<Constants>>().data
+                return _constants!!
+            } catch (ex: SerializationException) {
+                throw Exception("Ошибка сериализации. Проверьте сайт.")
+            }
         }
-        .build()
+        return _constants!!
+    }
 
-    // Latest
-    override fun latestUpdatesRequest(page: Int) = throw UnsupportedOperationException() // popularMangaRequest()
-    override fun fetchLatestUpdates(page: Int): Observable<MangasPage> {
-        if (csrfToken.isEmpty()) {
-            return client.newCall(popularMangaRequest(page))
-                .asObservableSuccess()
-                .flatMap { response ->
-                    // Obtain token
-                    val resBody = response.body.string()
-                    csrfToken = "_token\" content=\"(.*)\"".toRegex().find(resBody)!!.groups[1]!!.value
-                    return@flatMap fetchLatestMangaFromApi(page)
-                }
+    private fun checkForToken(chain: Interceptor.Chain): Response {
+        val req = chain.request().newBuilder()
+        if (chain.request().url.toString().contains("api.$apiDomain")) {
+            if (bearerToken.isNullOrBlank()) {
+                launchIO { getToken() }
+            }
+            req.apply {
+                addHeader("Authorization", bearerToken.orEmpty())
+            }
         }
-        return fetchLatestMangaFromApi(page)
+        return chain.proceed(req.build())
     }
 
-    private fun fetchLatestMangaFromApi(page: Int): Observable<MangasPage> {
-        return client.newCall(POST("$baseUrl/filterlist?dir=desc&sort=last_chapter_at&page=$page&chapters[min]=1", catalogHeaders()))
-            .asObservableSuccess()
-            .map { response ->
-                latestUpdatesParse(response)
+    @SuppressLint("SetJavaScriptEnabled")
+    @Suppress("NAME_SHADOWING")
+    private fun getToken() {
```

Comment:
consider changing it to [this implementation](https://github.com/keiyoushi/extensions-source/pull/3241/files#diff-b2fcea53275f77fafad564c13e2be399de15e208fd405ac4c00d3e0d2681af4bR94-R100)
as webview should be avoided where possible

@e-shl as [co-author](https://coauthoredby.netlify.app/)

Reply:
@e-shl implementation will not work because api endpoint expects bearer token, obtained by OAuth and saved in `localStorage['auth'].token` on log in.
So getting token from webview inevitable.
The problem is i don't know how to work with coroutines on android (sorry).

Token in `localStorage` are in format `{access_token: String, expires_in: Int, refresh_token: String, token_type: String}`
So i can propose saving `access_token`, `expires_in`, `timestamp` (login time) and `token_type` (incase site will change token type), and get new token when it expires 

I implemented `getConstants()` to prevent issues like this #2660 in future